### PR TITLE
Add OptionsFormatter class and update to allow auto-logging of Options

### PR DESCRIPTION
**Pull request by mrambacher**
This PR adds an OptionsFormatter to the ConfigOptions.  This class allows the options to be written and read in different formats.  Three formats initially supported for write are the existing Options-file properties format, the ";"-separated list of items, and writing the options to the log file.

By adding a LogOptionsFormatter, the system will be able to automatically log options to the file as they are added or removed.  Currently, this involves a secondary step of updating the Dump method which is sometimes missed.  Furthermore, some sub-classes may be skipped or not have their options printed at all.

Note that a later PR will use the LogOptionsFormatter to replace the code in Dump and GetPrintableOptions with the Options infrastructure and the LogOptionsFormatter

### CODE REVIEW

**Comment by udi-speedb** [comment]
@mrambacher - Please update when you have completed updating the branch. I will review only once the branch has all of the contents you intend to merge. Thanks

**Comment by udi-speedb** [comment]
Why is the title "WIP: ..."?

**Comment by udi-speedb** [comment]
@ofriedma, @mrambacher - Please complete the review and all required activities so this may be ready for merge. Thanks

**Comment by ofriedma** [comment]
> @ofriedma, @mrambacher - Please complete the review and all required activities so this may be ready for merge. Thanks

Will review after #648 review

**Comment by udi-speedb** [comment]
@ofriedma #648 is merged now. Please review this one. Thanks

File: options/options_formatter.cc [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+std::string LogOptionsFormatter::ToString(const std::string& prefix,
+                                          const OptionProperties& props) const {
+  std::string result;
+  if (!props.empty()) {
+    const auto& id = props.find(OptionTypeInfo::kIdPropName());
+    if (props.size() > 1) {
+      std::string spaces = "  ";
+      if (!prefix.empty()) {
+        // Indent by the number of "." in the prefix
+        for (int count = static_cast<int>(
```
Comment (ofriedma) [inline options/options_formatter.cc:286]: Is the casting necessary here?

**Review by ofriedma — CHANGES_REQUESTED** [review]

**Comment by ofriedma** [comment]
Please resolve conflicts  @mrambacher

**Comment by udi-speedb** [comment]
@ofriedma - Are you done reviewing the PR?

**Comment by ofriedma** [comment]
> @ofriedma - Are you done reviewing the PR?

waiting for a fix to my comments

File: options/options_formatter.cc (hunk 1 above)
Comment (mrambacher) [reply options/options_formatter.cc:286]: Changed to long and removed casting

**Comment by mrambacher** [comment]
> > @ofriedma - Are you done reviewing the PR?
> 
> waiting for a fix to my comments

Comments addressed.  Please review again

**Review by ofriedma — APPROVED** [review]

**Comment by ofriedma** [comment]
@mrambacher

**Review by ofriedma — CHANGES_REQUESTED** [review]
see comments

**Comment by mrambacher** [comment]
> @mrambacher There is a failure in the compile process

License check issue.  Believe it should be fixed or a bug in the license check for files that are not licensed.

**Review by ofriedma — APPROVED** [review]

**Review by ofriedma — CHANGES_REQUESTED** [review]
https://github.com/speedb-io/speedb/actions/runs/6839948302/job/18598695568

There is a compile issue again in the CI: 

In file included from ../options/options_formatter.cc:20:
../options/options_formatter_impl.h:56:15: error: 'virtual std::string rocksdb::DefaultOptionsFormatter::ToString(const string&, char, const std::vector<std::__cxx11::basic_string<char> >&) const' was hidden [-Werror=overloaded-virtual]
   56 |   std::string ToString(const std::string& prefix, char separator,
      |               ^~~~~~~~
../options/options_formatter_impl.h:92:15: note:   by 'virtual std::string rocksdb::PropertiesOptionsFormatter::ToString(const string&, const rocksdb::OptionProperties&) const'
   92 |   std::string ToString(const std::string& prefix,
      |               ^~~~~~~~
cc1plus: all warnings being treated as errors
ninja: job failed: ccache /usr/bin/c++ -DBZIP2 -DGFLAGS=1 -DGFLAGS_IS_A_DLL=0 -DHAVE_PCLMUL -DHAVE_SSE42 -DLZ4 -DOS_LINUX -DROCKSDB_FALLOCATE_PRESENT -DROCKSDB_LIB_IO_POSIX -DROCKSDB_MALLOC_USABLE_SIZE -DROCKSDB_NO_DYNAMIC_EXTENSION -DROCKSDB_PLATFORM_POSIX -DROCKSDB_PORTABLE -DROCKSDB_RANGESYNC_PRESENT -DROCKSDB_SCHED_GETCPU_PRESENT -DSNAPPY -DZLIB -DZSTD -I../ -I../include -isystem ../third-party/gtest-1.8.1/fused-src -W -Wextra -Wall -pthread -Wsign-compare -Wshadow -Wno-unused-parameter -Wno-unused-variable -Woverloaded-virtual -Wnon-virtual-dtor -Wno-missing-field-initializers -Wno-strict-aliasing -Wno-invalid-offsetof -fno-omit-frame-pointer -momit-leaf-frame-pointer -Werror -fno-builtin-memcmp -O3 -DNDEBUG -fno-rtti -fPIC -std=gnu++17 -MD -MT CMakeFiles/speedb.dir/options/options_parser.cc.o -MF CMakeFiles/speedb.dir/options/options_parser.cc.o.d -o CMakeFiles/speedb.dir/options/options_parser.cc.o -c ../options/options_parser.cc
In file included from ../options/options_parser.cc:33:
../options/options_formatter_impl.h:56:15: error: 'virtual std::string rocksdb::DefaultOptionsFormatter::ToString(const string&, char, const std::vector<std::__cxx11::basic_string<char> >&) const' was hidden [-Werror=overloaded-virtual]
   56 |   std::string ToString(const std::string& prefix, char separator,
      |               ^~~~~~~~
../options/options_formatter_impl.h:92:15: note:   by 'virtual std::string rocksdb::PropertiesOptionsFormatter::ToString(const string&, const rocksdb::OptionProperties&) const'
   92 |   std::string ToString(const std::string& prefix,
      |               ^~~~~~~~

**Review by ofriedma — DISMISSED** [review]

**Comment by ofriedma** [comment]
@mrambacher there is a compile error here:
../db/db_options_test.cc:64:12: error: local variable 'mutable_map' will be copied despite being returned by name [-Werror,-Wreturn-std-move]
    return mutable_map;
           ^~~~~~~~~~~
../db/db_options_test.cc:64:12: note: call 'std::move' explicitly to avoid copying
    return mutable_map;
           ^~~~~~~~~~~
           std::move(mutable_map)
../db/db_options_test.cc:76:12: error: local variable 'mutable_map' will be copied despite being returned by name [-Werror,-Wreturn-std-move]
    return mutable_map;
           ^~~~~~~~~~~
../db/db_options_test.cc:76:12: note: call 'std::move' explicitly to avoid copying
    return mutable_map;
           ^~~~~~~~~~~
           std::move(mutable_map)

**Review by ofriedma — APPROVED** [review]

**Pull request merged**